File: leetcode/tree/1120_Maximum_Average_Subtree/codes/dfs.py
from typing import Optional

# Definition for a binary tree node.
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right

class Solution:
    def maximumAverageSubtree(self, root: Optional[TreeNode]) -> float:
        self.max_avg = 0

        def dfs(node):
            if not node:
                return (0, 0)  
            l_sum, l_cnt = dfs(node.left)
            r_sum, r_cnt = dfs(node.right)

            total_sum = l_sum + r_sum + node.val
            total_count = l_cnt + r_cnt + 1
            avg = total_sum / total_count

            self.max_avg = max(self.max_avg, avg)
            return (total_sum, total_count)

        dfs(root)
        return self.max_avg

File: leetcode/tree/1120_Maximum_Average_Subtree/codes/dfs.c
/**
Definition for a binary tree node.
};
 */
struct TreeNode {
    int val;
    struct TreeNode *left;
    struct TreeNode *right;

struct Result {
    int sum;
    int count;
};

double maxAvg = 0;

struct Result dfs(struct TreeNode* node) {
    if (!node) return (struct Result){0, 0};

    struct Result left = dfs(node -> left);
    struct Result right = dfs(node -> right);

    int totalSum = left.sum + right.sum + node -> val;
    int totalCount = left.count + right.count + 1;

    double avg = (double)totalSum / totalCount;
    if (avg > maxAvg) maxAvg = avg;

    return (struct Result){totalSum, totalCount};
}

double maximumAverageSubtree(struct TreeNode* root) {
    maxAvg = 0;
    dfs(root);
    return maxAvg;
}
